### PR TITLE
LPS-119066 [Page Audit] New Resource Command that retrieve the list of LayoutReportsIssues given a URL.

### PR DESCRIPTION
**Pull request by cgoncas**
**Motivation**

We need to call the PageSpeed API in order to retrieve the list of SEO and Accessibility issues of aPage.

**Proposed Solution**

* Create a[ Resource Command](https://github.com/liferay-tango/liferay-portal/blob/0f9522eaaa0d3701e903e55c334826eead3716d9/modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java) that retrieves the number of SEO and Accessibility issues of a given CanonicalURL.

* The format of the JSON that returns the new Resource Command will be:

The format returned to the frontend will be:
```

{
      "layoutReportsIssues":[
         {
            "total":"2",
            "title":"Accessibily"
            "key":"accessibility"
         },
         {
            "total":"2",
            "title":"SEO",
            "key":"seo"
         }
      ]
}
```

**How to test it**

No Unit/Integration tests have been added to this PR

**_To test it manually_**

Call the new MVC Command:

* http://localhost:8080/web/guest/home?p_p_id=com_liferay_layout_reports_web_internal_portlet_LayoutReportsPortlet&p_p_lifecycle=2&p_p_state=normal&p_p_mode=view&p_p_resource_id=%2Flayout_reports%2Fget_layout_reports_issues&p_p_cacheability=cacheLevelPage&_com_liferay_layout_reports_web_internal_portlet_LayoutReportsPortlet_groupId=20125&_com_liferay_layout_reports_web_internal_portlet_LayoutReportsPortlet_canonicalURL=https%3A%2F%2Fliferay.com%2F
_Replacing the groupId with your groupId and the canonicalURL with the canonicalURL you want to check_

I have split the creation of the POJO to the PR https://github.com/liferay-tango/liferay-portal/pull/1081, so we can receive Brian's feedback about the names sooner.

### CODE REVIEW

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+		Locale locale = _portal.getLocale(resourceRequest);
+
+		try {
+			long groupId = ParamUtil.getLong(resourceRequest, "groupId");
+
+			Group group = _groupLocalService.fetchGroup(groupId);
+
+			if (group == null) {
+				_log.error("No API Key configured for the site " + groupId);
```
Comment (4lejandrito) [inline modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java:92]: I think this message is misleading. The problem is that the group does not exist. 

I tried calling the command without setting an API Key and it works fine (I guess with the restriction of no more than 1 or 2 requests per second).

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+				JSONPortletResponseUtil.writeJSON(
+					resourceRequest, resourceResponse,
+					JSONUtil.put(
+						"error",
+						"problems retrieving the API Key for the site " +
```
Comment (4lejandrito) [inline modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java:98]: Same with this one?

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+			String canonicalURL = ParamUtil.getString(
+				resourceRequest, "canonicalURL");
+
+			List<LayoutReportsIssue> layoutReportsIssues =
+				layoutReportsDataProvider.getLayoutReportsIssues(canonicalURL);
+
+			Stream<LayoutReportsIssue> stream = layoutReportsIssues.stream();
+
+			JSONPortletResponseUtil.writeJSON(
```
Comment (4lejandrito) [inline modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java:117]: If someone malicious discovers this endpoint, could they can get unlimited free API requests?

Should we check for some kind of permissions before calling the API?

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java:92]: Hey @4lejandrito !!

Yeah, we can call the service without APIKey, but we have decided that this is NOT OK for the client, and the client will need a configured APIKey in order to call the PageSpeed API:

**Given** the Page Audit Panel
**When** PageSpeed's API Key is not configured and the user has no permissions for accessing to the site/system/instance settings
**Then** a message and an image show: Check issues that impact on your page's accessibility and SEO. Connect with PageSpeed from Instance or System Site settings > Pages > Google PageSpeed.

But you are totally right about the message, I will change it :=

Thanks!!!

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java (hunk 3 above)
Comment (cgoncas) [reply modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java:117]: The problem with a permission, is that right now, you just need to have the view permission to see this statistics :(, probably we will add a new permission different that view/edit for it.

For now we have decided that the edit permission will be needed to call the API, but probably a more specific one will be created.

I will add the restriction for edit permission in LPS-119068, and the update this PR.

Thanks!!!

Thanks!!!

File: modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java (hunk 2 above)
Comment (cgoncas) [reply modules/apps/layout/layout-reports-web/src/main/java/com/liferay/layout/reports/web/internal/portlet/action/GetLayoutReportsIssuesMVCResourceCommand.java]: Yeahh!!!

**Comment by cgoncas** [comment]
Hey @4lejandrito !!

I will wait until https://github.com/liferay-tango/liferay-portal/pull/1075 is merged in order to change the view permission for edit permission.

Until that, I have send the POJO in the PR: https://github.com/liferay-tango/liferay-portal/pull/1081, so only the ResourceCommand is on hold.

Thanks!!!

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:forward

**Pull request closed without merging**